arrow-fx-coroutines/src/test/kotlin/arrow/fx/coroutines/parMapN/ParMap3Test.kt

```diff
@@ -0,0 +1,158 @@
+package arrow.fx.coroutines.parMapN
+
+import arrow.core.Either
+import arrow.fx.coroutines.ArrowFxSpec
+import arrow.fx.coroutines.Atomic
+import arrow.fx.coroutines.ExitCase
+import arrow.fx.coroutines.ForkAndForget
+import arrow.fx.coroutines.NamedThreadFactory
+import arrow.fx.coroutines.Promise
+import arrow.fx.coroutines.Resource
+import arrow.fx.coroutines.Semaphore
+import arrow.fx.coroutines.evalOn
+import arrow.fx.coroutines.guaranteeCase
+import arrow.fx.coroutines.never
+import arrow.fx.coroutines.parMapN
+import arrow.fx.coroutines.single
+import arrow.fx.coroutines.singleThreadName
+import arrow.fx.coroutines.suspend
+import arrow.fx.coroutines.threadName
+import arrow.fx.coroutines.throwable
+import io.kotest.matchers.shouldBe
+import io.kotest.property.Arb
+import io.kotest.property.arbitrary.element
+import io.kotest.property.arbitrary.int
+import io.kotest.property.arbitrary.string
+import java.util.concurrent.Executors
+
+class ParMap3Test : ArrowFxSpec(spec = {
+  "parMapN 3 returns to original context" {
+    val mapCtxName = "parMap3"
+    val mapCtx = Resource.fromExecutor { Executors.newFixedThreadPool(3, NamedThreadFactory { mapCtxName }) }
+
+    checkAll {
+      single.zip(mapCtx).use { (_single, _mapCtx) ->
+        evalOn(_single) {
+          threadName() shouldBe singleThreadName
+
+          val (s1, s2, s3) = parMapN(_mapCtx, threadName, threadName, threadName) { a, b, c -> Triple(a, b, c) }
+
+          s1 shouldBe mapCtxName
+          s2 shouldBe mapCtxName
+          s3 shouldBe mapCtxName
+          threadName() shouldBe singleThreadName
+        }
+      }
+    }
+  }
+
+  "parMapN 3 returns to original context on failure" {
+    val mapCtxName = "parMap3"
+    val mapCtx = Resource.fromExecutor { Executors.newFixedThreadPool(3, NamedThreadFactory { mapCtxName }) }
+
+    checkAll(Arb.int(1..3), Arb.throwable()) { choose, e ->
+      single.zip(mapCtx).use { (_single, _mapCtx) ->
+        evalOn(_single) {
+          threadName() shouldBe singleThreadName
+
+          Either.catch {
+            when (choose) {
+              1 -> parMapN(_mapCtx, suspend { e.suspend() }, suspend { never<Nothing>() }, suspend { never<Nothing>() }) { _, _, _ -> Unit }
+              2 -> parMapN(_mapCtx, suspend { never<Nothing>() }, suspend { e.suspend() }, suspend { never<Nothing>() }) { _, _, _ -> Unit }
+              else -> parMapN(_mapCtx, suspend { never<Nothing>() }, suspend { never<Nothing>() }, suspend { e.suspend() }) { _, _, _ -> Unit }
+            }
+          } shouldBe Either.Left(e)
+
+          threadName() shouldBe singleThreadName
+        }
+      }
+    }
+  }
+
+  "parMapN 3 runs in parallel" {
+    checkAll(Arb.int(), Arb.int(), Arb.int()) { a, b, c ->
+      val r = Atomic("")
+      val modifyGate1 = Promise<Int>()
+      val modifyGate2 = Promise<Int>()
```

Comment:
```suggestion
      val modifyGate1 = Promise<Unit>()
      val modifyGate2 = Promise<Unit>()
```

Reverting change to make similar examples for tests because it's necessary to adapt more sentences.